File: app/src/main/java/com/intoverflown/pos/ui/registerbranch/NewBranchActivity.java
package com.intoverflown.pos.ui.registerbranch;

import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.android.volley.DefaultRetryPolicy;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonArrayRequest;
import com.android.volley.toolbox.Volley;
import com.intoverflown.pos.MainActivity;
import com.intoverflown.pos.databinding.ActivityNewBranchBinding;
import com.intoverflown.pos.utils.Constants;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class NewBranchActivity extends AppCompatActivity {

    private ActivityNewBranchBinding binding;

    public SharedPreferences preferences;
    public SharedPreferences.Editor editor;
    public String SHARED_PREF_NAME = "pos_pref";
    public String KEY_ID = "Id";
    public String KEY_TOKEN = "Token";
    public String MERCHANT_ID = "merchantId";

    String uid;
    String token;
    Integer merchantId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityNewBranchBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.upBtn.setOnClickListener(v -> {
            Intent i = new Intent(NewBranchActivity.this, MainActivity.class);
            i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(i);
        });

        preferences = this.getSharedPreferences(SHARED_PREF_NAME, MODE_PRIVATE);

        uid = preferences.getString(KEY_ID, "Id");
        token = preferences.getString(KEY_TOKEN, "Token");
        merchantId = Integer.valueOf(preferences.getString(MERCHANT_ID, "merchantId"));

        Log.d("uid new branch", uid);
        Log.d("token new branch", token);

        binding.regMerchantBtn.setOnClickListener(v -> {
            String url = Constants.BASE_URL + "MerchantBranch/Create";
            postNewMerchant(url);
        });
    }

    private void postNewMerchant(String url) {
        RequestQueue queue = Volley.newRequestQueue(getApplicationContext());
        JSONArray array = new JSONArray();
        JSONObject jsonObjects = new JSONObject();

        String region = binding.regMerchantRegion.getText().toString().trim();
        String address = binding.regMerchantAddress.getText().toString().trim();
        String branchName = binding.regMerchantMerchantBranch.getText().toString().trim();
        String contactPerson = binding.regMerchantMerchantName.getText().toString().trim();
        String phone = binding.regMerchantPhone.getText().toString().trim();

        if (region.isEmpty() && address.isEmpty() &&
                branchName.isEmpty() && contactPerson.isEmpty() && phone.isEmpty()) {
            Toast.makeText(this, "Enter all fields!!", Toast.LENGTH_SHORT).show();
        } else {
            ProgressDialog progressDialog = new ProgressDialog(NewBranchActivity.this);
            progressDialog.setMessage("Creating new branch merchant!");
            progressDialog.show();

            try {
                jsonObjects.put("MerchantId", merchantId);
                jsonObjects.put("RegionId", region);
                jsonObjects.put("BranchName", branchName);
                jsonObjects.put("Address", address);
                jsonObjects.put("Phone", phone);
                jsonObjects.put("ContactPerson", contactPerson);
                jsonObjects.put("CreatedById", uid);

                array = new JSONArray("["+jsonObjects.toString()+"]");
                Log.d("Post", array.toString());
            } catch (final JSONException e) {
                e.printStackTrace();
            }

            JsonArrayRequest  jsonObjectRequest = new JsonArrayRequest(com.android.volley.Request.Method.POST,
                    url, array, response -> {
                try {
                    Log.d("connection", response.toString());
                    progressDialog.dismiss();

                    Log.d("response", response.toString());

                    for(int i=0; i < response.length(); i++) {
                        JSONObject jsonObj = response.getJSONObject(i);
                        Log.d("uid merchant branch", jsonObj.optString("uniqueId"));

                        // store id
                        editor = preferences.edit();
                        editor.putString("merchantBranchId", jsonObj.optString("uniqueId"));
                        editor.apply();
                    }

                    Toast.makeText(this, "Created Successfully!", Toast.LENGTH_SHORT).show();

                    intentBackHome();
                } catch (Exception e) {
                    Toast.makeText(this, "Error occurred\nCheck logs!", Toast.LENGTH_LONG).show();
                    Log.i("new branch", Log.getStackTraceString(e));
                }
            }, error -> {
                progressDialog.dismiss();
                Log.e("error", error.toString());
                Toast.makeText(NewBranchActivity.this, "Failed to create branch!", Toast.LENGTH_SHORT).show();
            }) {
                @Override
                public Map<String, String> getHeaders() {
                    Map<String, String> params = new HashMap<String, String>();
                    params.put("Content-Type", "application/json");
                    params.put("Authorization", "Bearer " + token);
                    return params;
                }

                @Override
                public String getBodyContentType() {
                    return "application/json";
                }
            };

            jsonObjectRequest.setRetryPolicy(new DefaultRetryPolicy(60000,
                    DefaultRetryPolicy.DEFAULT_MAX_RETRIES, DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));

            queue.add(jsonObjectRequest);
        }
    }

    private void intentBackHome() {
        Intent j = new Intent(NewBranchActivity.this, MainActivity.class);
        j.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(j);
        finish();
    }
}

File: app/src/main/java/com/intoverflown/pos/ui/registerbranch/NewBranchActivity.kt
package com.intoverflown.pos.ui.registerbranch

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.android.volley.AuthFailureError
import com.android.volley.DefaultRetryPolicy
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.intoverflown.pos.MainActivity
import com.intoverflown.pos.api_response.LoginResponse
import com.intoverflown.pos.data.SharedPreferenceManager
import com.intoverflown.pos.databinding.ActivityNewBranchBinding
import com.intoverflown.pos.ui.login.LoginActivity
import com.intoverflown.pos.utils.Constants
import org.json.JSONException
import org.json.JSONObject


class NewBranchActivity : AppCompatActivity() {

    private var binding: ActivityNewBranchBinding? = null
    private val context : SharedPreferenceManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityNewBranchBinding.inflate(layoutInflater)
        setContentView(binding!!.root)

        binding!!.upBtn.setOnClickListener {
            val x = Intent(this, MainActivity::class.java)
            x.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            startActivity(x)
        }
        /** Start*/
        val sharedPreferences = this.getSharedPreferences("volleyLogin", Context.MODE_PRIVATE)
        val usr = LoginResponse(
            sharedPreferences.getString("Id", "Id"),
            sharedPreferences.getString("FirstName", "FirstName"),
            sharedPreferences.getString("LastName", "LastName"),
            sharedPreferences.getString("Username", ""),
            sharedPreferences.getString("Role", "Role"),
            sharedPreferences.getString("Token", "Token")
        )
        // Store the user in shared pref
        SharedPreferenceManager.getInstance(applicationContext).getUser(usr)
        Log.d("id", usr.Id.toString())
        Log.d("token", usr.Token.toString())
        /** End*/

//        val sharedPreferences = this.getSharedPreferences("volleyLogin", Context.MODE_PRIVATE)
//
//
//        val uid = SharedPreferenceManager.getInstance(this@NewBranchActivity).getUser().Id
//        val token = SharedPreferenceManager.getInstance(this@NewBranchActivity).getUserTest()
//        val token = SharedPreferenceManager.getInstance(applicationContext).getTestToken()
//        Log.d("uid", uid)
//        Log.d("token pref", token)

        // post new merchant branch
        binding!!.regMerchantBtn.setOnClickListener {
            val url = Constants.BASE_URL + "MerchantBranch/Create"
//            postNewMerchant(url, uid.toString(), token.toString())
        }
    }

    private fun postNewMerchant(url: String, uid: String, token: String) {
        val queue = Volley.newRequestQueue(applicationContext)
        val jsonObject = JSONObject()

        val region = binding!!.regMerchantRegion.text.toString().trim()
        val branchName = binding!!.regMerchantMerchantBranch.text.toString().trim()
        val address = binding!!.regMerchantAddress.text.toString().trim()
        val phone = binding!!.regMerchantPhone.text.toString().trim()
        val contactPerson = binding!!.regMerchantMerchantName.text.toString().trim()

        if (region.isNotEmpty() && branchName.isNotEmpty() && address.isNotEmpty()
            && phone.isNotEmpty() && contactPerson.isNotEmpty()) {
            try {
                jsonObject.put("RegionId", region)
                jsonObject.put("BranchName", branchName)
                jsonObject.put("Address", address)
                jsonObject.put("Phone", phone)
                jsonObject.put("ContactPerson", contactPerson)
                jsonObject.put("CreatedById", uid)

                Log.d("Post", jsonObject.toString())
            } catch (e: JSONException) {
                e.printStackTrace()
            }

            val jsonObjectRequest: JsonObjectRequest = object : JsonObjectRequest(
                Method.POST, url, jsonObject, Response.Listener { response ->
                    Log.d("Post ", response.toString())
                    Toast.makeText(applicationContext, "Successful", Toast.LENGTH_SHORT).show()

                    val h = Intent(this, MainActivity::class.java)
                    h.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                    startActivity(h)

                }, Response.ErrorListener { error ->
                    Log.e("error", error.toString())
                    Toast.makeText(applicationContext, "Registration failed!", Toast.LENGTH_SHORT)
                        .show()
                }) {
                @Throws(AuthFailureError::class)
                override fun getHeaders(): Map<String, String> {
                    val params: MutableMap<String, String> = HashMap()
                    params["Authorization"] = "Bearer $token"
                    params["Content-Type"] = "application/json"
                    return params
                }

                override fun getBodyContentType(): String {
                    return "application/json"
                }
            }
            jsonObjectRequest.retryPolicy = DefaultRetryPolicy(
                60000,
                DefaultRetryPolicy.DEFAULT_MAX_RETRIES,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT
            )
            queue.add(jsonObjectRequest)
        } else {
            Toast.makeText(applicationContext, "Fill all details!", Toast.LENGTH_SHORT).show()
        }
    }
}